File: engine/aurora/sequence.py
"""Sequence Engine — frekans + süre + sıra + tekrar → akustik protokol (stimulus fingerprint)."""
from __future__ import annotations

import hashlib
import json
import random
from dataclasses import asdict

from .models import Step, Stimulus

FIBONACCI_MIN = (5, 8, 13)
POPULAR = (174, 285, 396, 417, 432, 528, 639, 741, 852, 963)


def fingerprint(steps: list[Step], gap_s: float, repetitions: int, waveform: str, amplitude: float) -> str:
    payload = json.dumps({"steps": [asdict(s) for s in steps], "gap": gap_s, "rep": repetitions,
                          "wave": waveform, "amp": amplitude}, sort_keys=True)
    return "STM-" + hashlib.sha1(payload.encode()).hexdigest()[:10].upper()


def build(freqs: list[float], minutes: list[float], gap_s: float = 0.5, repetitions: int = 1,
          waveform: str = "sine", amplitude: float = 0.15) -> Stimulus:
    if len(freqs) != len(minutes):
        raise ValueError("freqs and minutes must have equal length")
    if any(m <= 0 for m in minutes):
        raise ValueError("durations must be positive")
    steps = [Step(float(f), float(m) * 60.0) for f, m in zip(freqs, minutes)]
    return Stimulus(fingerprint(steps, gap_s, repetitions, waveform, amplitude), tuple(steps),
                    gap_s, repetitions, waveform, amplitude)


def protocol_set(freqs: list[float], minutes: list[float] = list(FIBONACCI_MIN), seed: int = 0,
                 lo: float = 100.0, hi: float = 1000.0) -> dict[str, Stimulus]:
    """Karşılaştırma seti: A ileri, B ters süre, C tek frekans, D rastgele frekans, E sessizlik."""
    rng = random.Random(seed)
    n = len(freqs)
    mins = list(minutes)[:n] if len(minutes) >= n else [minutes[i % len(minutes)] for i in range(n)]
    total = sum(mins)
    return {
        "A_forward": build(freqs, mins),
        "B_reversed_durations": build(freqs, list(reversed(mins))),
        "C_single": build([freqs[0]], [total]),
        "D_random": build([round(rng.uniform(lo, hi), 2) for _ in range(n)], mins),
        "E_silence": build([0.0] * n, mins),
    }


def permutations_of(freqs: list[float], minutes: list[float]) -> list[Stimulus]:
    from itertools import permutations
    return [build(list(p), minutes) for p in permutations(freqs)]


def to_json(stim: Stimulus) -> dict:
    return {"stimulus_id": stim.stimulus_id, "sequence": [s.hz for s in stim.steps],
            "durations": [s.duration_s for s in stim.steps], "gap_s": stim.gap_s,
            "repetitions": stim.repetitions, "waveform": stim.waveform, "amplitude": stim.amplitude,
            "sample_rate": stim.sample_rate, "total_seconds": stim.total_seconds}

File: app/src/core/sequence.ts
/** Sequence Engine — frekans + süre + sıra + tekrar → stimulus fingerprint */
import type { Stimulus, Step, WaveformKind } from "../data/types";

const hash = (s: string): string => {
  let h = 0x811c9dc5;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 0x01000193) >>> 0;
  }
  return h.toString(16).toUpperCase().padStart(8, "0");
};

export const fingerprint = (steps: Step[], gapS: number, repetitions: number, waveform: WaveformKind, amplitude: number): string =>
  "STM-" + hash(JSON.stringify({ steps, gapS, repetitions, waveform, amplitude }));

export const totalSeconds = (s: Pick<Stimulus, "steps" | "gapS" | "repetitions">): number =>
  (s.steps.reduce((a, x) => a + x.durationS, 0) + s.gapS * Math.max(0, s.steps.length - 1)) * s.repetitions;

export const buildStimulus = (
  freqs: number[],
  minutes: number[],
  opts: Partial<Pick<Stimulus, "gapS" | "repetitions" | "waveform" | "amplitude" | "name">> = {},
): Stimulus => {
  if (freqs.length !== minutes.length) throw new Error("freqs and minutes must have equal length");
  if (minutes.some((m) => m <= 0)) throw new Error("durations must be positive");
  const steps = freqs.map((hz, i) => ({ hz, durationS: minutes[i] * 60 }));
  const gapS = opts.gapS ?? 0.5;
  const repetitions = opts.repetitions ?? 1;
  const waveform = opts.waveform ?? "sine";
  const amplitude = Math.min(opts.amplitude ?? 0.15, 0.2);
  return { id: fingerprint(steps, gapS, repetitions, waveform, amplitude), name: opts.name ?? "", steps, gapS, repetitions, waveform, amplitude };
};

const mulberry = (seed: number) => () => {
  let t = (seed += 0x6d2b79f5);
  t = Math.imul(t ^ (t >>> 15), t | 1);
  t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
  return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
};

export const protocolSet = (freqs: number[], minutes = [5, 8, 13], seed = 0, lo = 100, hi = 1000): Record<string, Stimulus> => {
  const rng = mulberry(seed);
  const mins = freqs.map((_, i) => minutes[i % minutes.length]);
  const total = mins.reduce((a, b) => a + b, 0);
  return {
    A_forward: buildStimulus(freqs, mins, { name: "A" }),
    B_reversed: buildStimulus(freqs, [...mins].reverse(), { name: "B" }),
    C_single: buildStimulus([freqs[0]], [total], { name: "C" }),
    D_random: buildStimulus(freqs.map(() => Math.round((lo + rng() * (hi - lo)) * 100) / 100), mins, { name: "D" }),
    E_silence: buildStimulus(freqs.map(() => 0), mins, { name: "E" }),
  };
};
